### ISSUE TITLE
`nextstrain update conda` in 7.0.0 can error with "Invalid version"

### ISSUE THREAD

**Issue by tsibley** [nextstrain/cli#285]
Running the standalone version of 7.0.0 I get:

```
$ nextstrain update conda
Checking for newer versions of Nextstrain CLI…

nextstrain-cli is up to date!

Updating conda runtime…
Traceback (most recent call last):
  File "runpy", line 196, in _run_module_as_main
  File "runpy", line 86, in _run_code
  File "nextstrain.cli.__main__", line 55, in <module>
  File "nextstrain.cli.__main__", line 19, in main
  File "nextstrain.cli", line 36, in run
  File "nextstrain.cli.command.update", line 54, in run
  File "nextstrain.cli.runner.conda", line 460, in update
  File "nextstrain.cli.runner.conda", line 552, in package_distribution
  File "nextstrain.cli.runner.conda", line 595, in latest_package_label_version
  File "nextstrain.cli.runner.conda", line 595, in <lambda>
  File "packaging.version", line 52, in parse
  File "packaging.version", line 198, in __init__
packaging.version.InvalidVersion: Invalid version: '20221019T172207Z'
```

This also happens with a Pip-based install of 7.0.0, as long as `packaging` is new enough.

This happens because the `packaging` library dropped support for its `LegacyVersion` class in 22.0 (December 2022) ([changelog](https://packaging.pypa.io/en/latest/changelog.html)).

Why this didn't fail in CI for #280, I'm not sure. Possibly the code isn't covered by our CI tests.

**Comment by joverlee521** [comment]
Arg, sorry I missed this. When I was testing locally with the `.venv/` environment, the environment was pinned to `packaging` v21.3 since v22.0 dropped support for Python3.6.

**Comment by tsibley** [comment]
No worries at all!

I'm a bit surprised `nextstrain setup conda` didn't catch this in CI, but I guess this code path isn't taken by `setup` and CI doesn't do `nextstrain update conda`.

When library compat issues like this come up, my thought process for a solution is generally to consider each of these in turn:

1. Pin?
2. Vendor?
3. Re-implement bare minimum?
4. Stop using the library by rearranging our needs?

I don't think we should pin `packaging <22` because that's not a sustainable solution. I don't think we should vendor `LegacyVersion` because it's fairly complex.
I think we could re-implement a bare minimum either as:

1. Catch `InvalidVersion` and fall back to a new `StringVersion` class which has the comparison semantics of a string plus always compares less than a PEP-440-compliant `Version` object. This is not the same as what `LegacyVersion` did, but I think it's good enough for our purposes.

2. Catch `InvalidVersion` and retry as `Version(f"0+{version}")`. This relies on the invalid `version` being a [valid PEP-440 local segment](https://peps.python.org/pep-0440/#local-version-segments) and [reusing their comparison semantics](https://peps.python.org/pep-0440/#local-version-identifiers). It seems likely that versions would be valid local segments since local segments can contain `[A-Za-z0-9._-]` (with `_` and `-` being normalized to `.`, which is then split on for comparison by parts).

**Comment by tsibley** [comment]
The first possible solution seems less hacky in some ways. But the second will more correctly compare versions like `1.0.0-foo`.

It's worth noting that we only use this for `micromamba` (from conda-forge) and `nextstrain-base` (from our channel). (Ignoring the possibility that different channels and packages can be used for the runtime package; we don't have to support that for anything but dev/testing.)

So if Micromamba's versions are PEP-440 compliant, then a simple string comparison for our `nextstrain-base` versions is fine.

**Comment by tsibley** [comment]
Actually, we don't even use it for `micromamba` in practice because we pin that to a known version (`1.0.0`).

**Comment by tsibley** [comment]
(But all of `micromamba`'s versions so far are PEP-440 compliant.)

**Comment by tsibley** [comment]
Is this terrible?

```diff
diff --git a/nextstrain/cli/runner/conda.py b/nextstrain/cli/runner/conda.py
index 502a950..7ffcecd 100644
--- a/nextstrain/cli/runner/conda.py
+++ b/nextstrain/cli/runner/conda.py
@@ -46,7 +46,7 @@ import shutil
 import subprocess
 import tarfile
 import traceback
-from packaging.version import parse as parse_version
+from packaging.version import Version, InvalidVersion
 from pathlib import Path, PurePosixPath
 from typing import Iterable, NamedTuple, Optional
 from urllib.parse import urljoin, quote as urlquote
@@ -596,6 +596,14 @@ def latest_package_label_version(channel: str, package: str, label: str) -> Opti
     return latest_file.get("version")
 
 
+def parse_version(version: str) -> Version:
+    try:
+        return Version(version)
+    except InvalidVersion:
+        version_as_local_segment = re.sub(r'[^A-Za-z0-9._-]', '.', version)
+        return Version(f"0+{version_as_local_segment}")
+
+
 class PackageSpec(NamedTuple):
     name: str
     version_spec: Optional[str] = None
```

**Comment by joverlee521** [comment]
Eh, I think it's acceptable for our internal versions. 

Maybe need to adjust the regex? 
Just reading the PEP-440 [local version identifier](https://peps.python.org/pep-0440/#local-version-identifiers) docs:
```
 local version labels MUST be limited to the following set of permitted characters:

    ASCII letters ([a-zA-Z])
    ASCII digits ([0-9])
    periods (.)

Local version labels MUST start and end with an ASCII letter or digit.
```

**Comment by tsibley** [comment]
`_` and `-` are allowed too but are [normalized](https://peps.python.org/pep-0440/#local-version-segments) to `.`:

> With a local version, in addition to the use of `.` as a separator of segments, the use of `-` and `_` is also acceptable. The normal form is using the `.` character. This allows versions such as `1.0+ubuntu-1` to be normalized to `1.0+ubuntu.1`.

but yeah, starting/ending with an ASCII letter/digit is something the regex doesn't do.

**Comment by joverlee521** [comment]
Ah, consecutive `.`, `_`, `-` aren't allowed either, since that would result in an "empty" segment.

**Comment by joverlee521** [comment]
Should we only use the local version segment only when the version string matches our current date-based versions and just re-raise the `InvalidVersion` error if we encounter something unknown?

**Comment by tsibley** [comment]
https://github.com/nextstrain/cli/compare/trs/conda/broken-update is the updated version

**Comment by tsibley** [comment]
(Putting this down for the day/week. Will return to it Tuesday unless you want to pick it up!)

**Comment by tsibley** [comment]
Fixed in 7.0.1, to be released soon.
